pkg/controllers/user/approuter/ingress_controller.go

```diff
@@ -0,0 +1,175 @@
+package approuter
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/rancher/rancher/pkg/ticker"
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/extensions/v1beta1"
+	"github.com/sirupsen/logrus"
+	extensionsv1beta1 "k8s.io/api/extensions/v1beta1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"strings"
+)
+
+const (
+	annotationHostname        = "rdns.cattle.io/hostname"
+	annotationIngressClass    = "kubernetes.io/ingress.class"
+	ingressClassNginx         = "nginx"
+	refreshIngressHostnameKey = "_refreshRDNSHostname_"
+	RdnsIPDomain              = "lb.rancher.cloud"
+)
+
+var (
+	renewInterval = 24 * time.Hour
+	ips           []string
+)
+
+type Controller struct {
+	ctx                    context.Context
+	ingressInterface       v1beta1.IngressInterface
+	ingressLister          v1beta1.IngressLister
+	managementSecretLister v1.SecretLister
+	clusterName            string
+	dnsClient              *Client
+}
+
+func isGeneratedDomain(obj *extensionsv1beta1.Ingress, host, domain string) bool {
+	parts := strings.Split(host, ".")
+	return strings.HasSuffix(host, "."+domain) && len(parts) == 6 && parts[1] == obj.Namespace
+}
+
+func (c *Controller) sync(key string, obj *extensionsv1beta1.Ingress) error {
+	//will not do clean if domain setting is cleaned
+	serverURL := settings.RDNSServerBaseURL.Get()
+	if serverURL == "" {
+		logrus.Warnf("settings.baseRDNSServerURL is not set, dns name might not be reachable")
+	}
+
+	for _, status := range obj.Status.LoadBalancer.Ingress {
+		if status.IP != "" {
+			ips = append(ips, status.IP)
+		}
+	}
+
+	if serverURL != c.dnsClient.GetBaseURL() {
```

Comment:
we lock in GetBaseURL only, so there are chances for SetBaseURL to be overridden by 2 clients. Can we just:

* call SetBaseURL
* lock in SetBaseURL
* do the check if domains are equal there, and then make an assignment

so the entire logic will be locked.